src/django_pg_migration_tools/management/commands/migrate_with_timeouts.py

```diff
@@ -54,13 +210,137 @@ def _validate(self) -> None:
+                ):
+                    super().handle(*django_migrate_args, **django_migrate_kwargs)
+                    migration_applied = True
+            except timeouts.DBLockTimeoutError as exc:
+                self.lock_timeouts += 1
+                self._wait(
+                    exp=self.lock_timeout_retry_exp,
+                    min_wait=self.lock_timeout_retry_min_wait,
+                    max_wait=self.lock_timeout_retry_max_wait,
+                    current_attempt=self.lock_timeouts,
+                )
+                self._attempt_callback(exc)
+            except timeouts.DBStatementTimeoutError as exc:
+                self.statement_timeouts += 1
+                self._wait(
+                    exp=self.statement_timeout_retry_exp,
+                    min_wait=self.statement_timeout_retry_min_wait,
+                    max_wait=self.statement_timeout_retry_max_wait,
+                    current_attempt=self.statement_timeouts,
+                )
+                self._attempt_callback(exc)
+
+        if migration_applied is False:
+            raise base.CommandError(
+                f"Please consider trying a longer retry configuration or "
+                f"investigate whether there were long-running transactions "
+                f"during the migration. "
+                f"There were {self.lock_timeouts} lock timeouts and "
+                f"{self.statement_timeouts} statement timeouts. This happened "
+                f"because --lock-timeout-max-retries was set to "
+                f"{self.lock_timeout_max_retries} and "
+                f"--statement-timeout-max-retries was set to "
+                f"{self.statement_timeout_max_retries}."
+            )
+
+    def _can_migrate(self) -> bool:
+        if self.lock_timeouts == 0 and self.statement_timeouts == 0:
+            # No timeouts implies this is the first time the migration will
+            # run.
+            return True
+
+        return bool(
+            (self.lock_timeouts <= self.lock_timeout_max_retries)
+            and (self.statement_timeouts <= self.statement_timeout_max_retries)
+        )
+
+    def _wait(
+        self,
+        exp: int,
+        min_wait: datetime.timedelta,
+        max_wait: datetime.timedelta,
+        current_attempt: int,
+    ) -> None:
+        if not self._can_migrate():
+            # No point waiting if we can't migrate.
+            return
+        try:
+            # current_attempt is an integer, but it is turned into a float here
+            # because a huge exponentiation in Python between integers
+            # **never** overflows. Instead, the CPU is left trying to calculate
+            # the result forever and it will eventually return a memory error
+            # instead. Which we absolutely do not want. Please see:
+            # https://docs.python.org/3.12/library/exceptions.html#OverflowError
+            result = exp ** (float(current_attempt))
```

Comment:
How many retries do we expect users to attempt? More than thousands???

Counterpoint: this code will cope with any number of retries, and with no particular downsides apart from the need to explain why the cast to float is done.

Reply:
Not in the thousands. Hopefully not even in the hundreds.
You can, of course, run into overflow errors in less than a thousand attempts for a high exponential value - but I also hope this to not be a common scenario. Giving that I am "hoping for many things" I agree that it's best to make the code work for any number of retries lest an arbitrary limiting number is incorrectly chosen and we have to go through a soft deprecation process.